tests/test_template_alias_base.cpp

```diff
@@ -0,0 +1,24 @@
+#include "pybind11_tests.h"
+
+#include <memory>
+
+template <class T>
+using DefaultAllocator = std::allocator<T>;
+
+template <template <class> class Allocator = DefaultAllocator>
+struct Base {};
+
+template <template <class> class Allocator = DefaultAllocator>
+struct S : public Base<Allocator> {};
+
+// this returns S<DefaultAllocator> even though we register
+// the type S<std::allocator>
+// MSVC and Clang on Windows failed here to detect the registered type
+S<> make_S() { return S<>{}; }
```

Comment:
This is basically stress-testing the compiler+library: can it figure out that `S<DefaultAllocator>` and `S<std::allocator>` are the same?

Question 1: Is it difficult to not have that stress-test baked into your downstream project? Either, be consistently avoiding the alias, or consistently using the alias.

Question 2: What happens if you patch internals.h similar to #4319?

If you decide to experiment more here, could you please not force push? That makes it more difficult for me to follow.

Reply:
> This is basically stress-testing the compiler+library: can it figure out that S<DefaultAllocator> and S<std::allocator> are the same?

Correct.

> Question 1: Is it difficult to not have that stress-test baked into your downstream project? Either, by consistently avoiding the alias, or consistently using the alias.

This is somewhat tricky in my case, especially across library/module interfaces. By consistently using one or the other, one has to hope a user does never use the other possible option of the alias.

> Question 2: What happens if you patch internals.h similar to https://github.com/pybind/pybind11/pull/4319?

Will try and push here.

> If you decide to experiment more here, could you please not force push? That makes it more difficult for me to follow.

Will do! I just force-pushed initially yesterday when I realized the initial test draft was not yet complete/buggy.

Reply:
Pushed patch :)

Reply:
@rwgk I don't understand your patch yet:
The comment says, GNU's `libstdc++` is good - all other implementations need the self-made hash.

Your patch says LLVM's `libc++` is not good - all other implementations can use the standard hashes.

Reply:
My test works on Linux with libstdc++.